### PR TITLE
GH-59: Multi-column field support (for cardinality 0)

### PR DESCRIPTION
**Pull request by upchuk**
Fixes GH-59.

### CODE REVIEW

**Comment by ademarco** [comment]
Tested locally by adding a link field to a RDF entity bundle, I've used the same mappings used on the test, I can confirm it works well. :+1:

**Review by voidtek — APPROVED** [review]
tested and approvated

File: rdf_entity.module [hunk 1]
```diff
@@ -76,6 +76,7 @@ function rdf_entity_form_field_storage_config_edit_form_alter(&$form, FormStateI
     '#description' => t('This field uses a Sparql backend. Please map the fields to their corresponding rdf properties.'),
     '#weight' => 99,
   ];
+  $form_state->set('rdf_mapping_columns', array_keys($schema['columns']));
```
Comment (claudiu-cristea) [inline rdf_entity.module:80]: I think we can simplify the whole logic:

1. In form alter we add `'#tree' => TRUE` in `$form['rdf_mapping']`.
1. We rename `$form['rdf_mapping'][$column]['predicate_' . $column]` to `$form['rdf_mapping'][$column]['predicate']`. Same for 'format'.
1. In `rdf_entity_field_storage_form_validate()`, we filter out empty predicates (anyway we don't need to store them):
    ```php
    $form_state->setValue('rdf_mapping', array_filter($form_state->getValue('rdf_mapping'), function (array $column): bool {
      return !empty($column['predicate']);
    }));
    ```
1. We check the non-empty predicate columns count `count($form_state->getValue('rdf_mapping'))`, together with the cardinality.
1. Of course `rdf_entity_form_alter_builder()` should be fixed too.

File: rdf_entity.module [hunk 2]
```diff
@@ -110,6 +111,38 @@ function rdf_entity_form_field_storage_config_edit_form_alter(&$form, FormStateI
...
+ * Validation callback for the field storage form.
+ *
+ * @param array $form
+ *   Form definition.
+ * @param \Drupal\Core\Form\FormStateInterface $form_state
+ *   Form state.
+ */
+function rdf_entity_field_storage_form_validate(array &$form, FormStateInterface $form_state) {
+  $cardinality = (int) $form_state->getValue('cardinality_number');
+  if ($cardinality < 2 && $form_state->getValue('cardinality') !== "-1") {
```
Comment (claudiu-cristea) [inline rdf_entity.module]: Use the `CARDINALITY_UNLIMITED` constant instead of -1.

File: src/Entity/RdfEntitySparqlStorage.php [hunk 3]
```diff
@@ -363,11 +363,11 @@ protected function processGraphResults($results, array $graph_ids): ?array {
             $column = $inbound_map['fields'][$predicate][$bundle]['column'];
             foreach ($field as $lang => $items) {
               $langcode_key = ($lang === $default_language) ? LanguageInterface::LANGCODE_DEFAULT : $lang;
-              foreach ($items as $item) {
+              foreach ($items as $delta => $item) {
                 $item = $this->fieldHandler->getInboundValue($this->getEntityTypeId(), $field_name, $item, $langcode_key, $column, $bundle);
 
                 if (!isset($return[$entity_id][$field_name][$langcode_key]) || !is_string($return[$entity_id][$field_name][$langcode_key])) {
-                  $return[$entity_id][$field_name][$langcode_key][][$column] = $item;
+                  $return[$entity_id][$field_name][$langcode_key][$delta][$column] = $item;
```
Comment (claudiu-cristea) [inline src/Entity/RdfEntitySparqlStorage.php:370]: I'm not sure about this change. Can you explain?

**Review by claudiu-cristea — COMMENTED** [review]
Some remarks and a question

File: src/Entity/RdfEntitySparqlStorage.php (hunk 3 above)
Comment (upchuk) [reply src/Entity/RdfEntitySparqlStorage.php:370]: This is the actual change that fixes the code. It adds together the values of multiple columns using the field delta. Without this, for each column that is mapped, a new field delta used to get created.

**Pull request merged**